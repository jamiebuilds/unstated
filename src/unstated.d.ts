import * as React from 'react';

export class Container<State extends object> {
  state: State;
  setState<K extends keyof State>(
    state:
      | ((prevState: Readonly<State>) => Pick<State, K> | State | null)
      | (Pick<State, K> | State | null),
    callback?: () => void
  ): Promise<void>;
  subscribe(fn: () => any): void;
  unsubscribe(fn: () => any): void;
}

export interface ContainerType<State extends object> {
  new(...args: any[]): Container<State>;
}

interface SubscribeProps {
  to: (ContainerType<any> | Container<any>)[];
  children(...instances: Container<any>[]): React.ReactNode;
}

export class Subscribe extends React.Component<SubscribeProps> { }

export interface ProviderProps {
  inject?: Container<any>[];
  children: React.ReactNode;
}

export const Provider: React.SFC<ProviderProps>;

export function useUnstated(...containers: (ContainerType<any> | Container<any>)[]): Container<any>[];
