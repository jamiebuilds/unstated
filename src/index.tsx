import React  from 'react'
import hoistNonReactStatic from 'hoist-non-react-statics'

type Listener = () => Promise<void>

export class Container<S extends object> {
  state: S
  _listeners: Listener[] = []

  constructor() {
    CONTAINER_DEBUG_CALLBACKS.forEach(cb => cb(this))
  }

  setState(
    updater?: Partial<S> | ((prevState: S) => Partial<S>),
    callback?: (state?: S) => void
  ): Promise<void> {
    let nextState: Partial<S>
    if (typeof updater === 'function') {
      nextState = (updater as (prevState: S) => Partial<S>)(this.state)
    } else {
      nextState = updater
    }

    if (nextState === null) {
      if (callback) callback(this.state)
      return Promise.resolve()
    }

    if (nextState && nextState !== this.state) {
      this.state = {...this.state, ...nextState}
    }
    
    let promises = this._listeners.map(listener => listener())
    return Promise.all(promises).then(() => {
      if (callback && typeof callback === 'function') {
        callback(this.state)
      }
    })
  }

  subscribe(fn: Listener) {
    this._listeners.push(fn);
  }

  unsubscribe(fn: Listener) {
    this._listeners = this._listeners.filter(f => f !== fn);
  }
}

interface ContainerClass<S extends object, TContainer extends Container<S> = Container<S>> {
  new (...args: any[]): TContainer
}

type ContainerType<S extends object> = Container<S> | ContainerClass<S>

type ContainersType = [ContainerType<object>, ...ContainerType<object>[]]

type ContainersMap = Map<ContainerClass<object>, Container<object>>

type MapContainersType<TContainers extends ContainersType> = {
  [K in keyof TContainers]: TContainers[K] extends ContainerClass<object, infer C> ? C : 
    TContainers[K] extends Container<object> ? TContainers[K] : any
}

type Containers<
  TContainers extends ContainerType<object> | ContainersType
> = TContainers extends ContainerClass<object, infer C> ? [C] : 
    TContainers extends Container<object> ? [TContainers] :
    TContainers extends ContainersType ? MapContainersType<TContainers> :
    any[]

export interface ISubscribeProps<TContainers extends ContainersType> {
  to: TContainers
  children: (...instances: Containers<TContainers>) => React.ReactNode
}

const Context = React.createContext<ContainersMap>(null)  // type ContainersMap

export class Subscribe<TContainers extends ContainersType> extends React.Component<ISubscribeProps<TContainers>> {
  state = {}
  _instances = []
  _unmounted = false

  _unsubscribe() {
    this._instances.forEach(container => {
      container.unsubscribe(this.onUpdate)
    })
  }

  _createInstances(
    ctx: ContainersMap,
    containers: TContainers
  ) {
    this._unsubscribe()

    if (!ctx) {
      throw new Error(
        'You must wrap your <Subscribe> components with a <Provider>'
      )
    }

    this._instances = containers.map(item => {
      let instance: ContainerType<object>
      if (typeof item === 'object' && item instanceof Container) {
        instance = item
      } else {
        instance = ctx.get(item)
        if (!instance) {
          instance = new item()
          ctx.set(item, instance)
        }
        instance.subscribe(this.onUpdate)
      }

      return instance
    })

    return this._instances as Containers<TContainers>
  }

  componentWillUnmount() {
    this._unmounted = true
    this._unsubscribe()
  }

  onUpdate: Listener = async () => {
    return new Promise(resolve => {
      if (!this._unmounted) {
        this.setState({}, resolve)
      } else {
        resolve()
      }
    })
  }

  render() {
    const { to, children } = this.props
    return (
      <Context.Consumer>
        {ctx => children.apply(null, this._createInstances(ctx, to))}
      </Context.Consumer>
    )
  }
}

interface IProviderProps {
  inject?: [Container<object>, ...Container<object>[]]
  children: React.ReactNode
}

export const Provider = ({inject, children}: IProviderProps) => {
  return (
    <Context.Consumer>
      {ctx => {
        let map = new Map(ctx)
  
        if (inject) {
          inject.forEach(instance => {
            map.set(instance.constructor as any, instance)
          })
        }
  
        return (
          <Context.Provider value={map}>
            {children}
          </Context.Provider>
        )
      }}
    </Context.Consumer>
  )
}

type IMapStateToProps<
  TContainers extends ContainerType<object> | ContainersType
> = (...containers: Containers<TContainers>) => object

export const unstated = <
  TContainers extends ContainerType<object> | ContainersType
>(containers: TContainers, mapStateToProps?: IMapStateToProps<TContainers>) => 
  <P extends object>(Component: React.ComponentType<P>) => {
    class UnstatedComponent extends React.Component<P> {
      render() {
        return (
          <Subscribe to={(Array.isArray(containers) ? containers: [containers]) as ContainersType}>{
            (...containers) => {
              let injectProps = {}
              if (mapStateToProps === undefined) {
                containers.forEach(c => {
                  let container = c.constructor.name
                  container = container.charAt(0).toLowerCase() + container.slice(1)

                  injectProps = {
                    ...injectProps,
                    [container]: c
                  }
                })
              } else {
                injectProps = mapStateToProps(...containers as any)
              }
              return <Component {...this.props} {...injectProps}/>
            }
          }</Subscribe>
        )
      }
    }

  // display name
  (UnstatedComponent as React.ComponentType).displayName =
    `Unstated(${Component.displayName || Component.name || 'Component'})`

  // Copy statics
  hoistNonReactStatic(UnstatedComponent, Component)

  return UnstatedComponent as React.ComponentType<any>
}

export default unstated

/* FOR DEBUG START */
type ContainerDebugCallback = (container: Container<object>) => void

let CONTAINER_DEBUG_CALLBACKS: ContainerDebugCallback[] = []

// If your name isn't Sindre, this is not for you.
// I might ruin your day suddenly if you depend on this without talking to me.
export function __SUPER_SECRET_CONTAINER_DEBUG_HOOK__(
  callback: ContainerDebugCallback
) {
  CONTAINER_DEBUG_CALLBACKS.push(callback)
}
/* FOR DEBUG END */
