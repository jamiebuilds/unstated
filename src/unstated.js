// @flow
import React, { type Node } from 'react';
import createReactContext from 'create-react-context';
import PropTypes from 'prop-types';
import defer from 'tickedoff';

type Listener = (cb?: () => void) => void;

const StateContext = createReactContext(null);

export class Container<State: {}> {
  state: State;
  _listeners: Array<Listener> = [];

  setState(
    updater: $Shape<State> | ((prevState: $Shape<State>) => $Shape<State>),
    callback?: () => void
  ) {
    defer(() => {
      let nextState;

      if (typeof updater === 'function') {
        nextState = updater(this.state);
      } else {
        nextState = updater;
      }

      if (nextState == null) {
        if (callback) callback();
        return;
      }

      this.state = Object.assign({}, this.state, nextState);

      let completed = 0;
      let total = this._listeners.length;

      this._listeners.forEach(fn => {
        if (!callback) {
          fn();
          return;
        }

        let safeCallback = callback;

        fn(() => {
          completed++;
          if (completed < total) {
            safeCallback();
          }
        });
      });
    });
  }

  subscribe(fn: Listener) {
    this._listeners.push(fn);
  }

  unsubscribe(fn: Listener) {
    this._listeners = this._listeners.filter(f => f !== fn);
  }
}

export type ContainerType = Container<Object>;
export type ContainersType = Array<Class<ContainerType> | ContainerType>;
export type ContainerMapType = Map<Class<ContainerType>, ContainerType>;

export type SubscribeProps<Containers: ContainersType> = {
  to: Containers,
  children: (
    ...instances: $TupleMap<Containers, <C>(Class<C> | C) => C>
  ) => Node
};

type SubscribeState = {};

const DUMMY_STATE = {};

export class Subscribe<Containers: ContainersType> extends React.Component<
  SubscribeProps<Containers>,
  SubscribeState
> {
  static propTypes = {
    to: PropTypes.array.isRequired,
    children: PropTypes.func.isRequired
  };

  state = {};
  instances: Array<ContainerType> = [];

  componentWillUnmount() {
    this._unsubscribe();
  }

  _unsubscribe() {
    this.instances.forEach(container => {
      container.unsubscribe(this.onUpdate);
    });
  }

  onUpdate: Listener = cb => {
    this.setState(DUMMY_STATE, cb);
  };

  _createInstances(
    map: ContainerMapType | null,
    containers: ContainersType
  ): Array<ContainerType> {
    this._unsubscribe();

    if (map === null) {
      throw new Error(
        'You must wrap your <Subscribe> components with a <Provider>'
      );
    }

    let safeMap = map;
    let instances = containers.map(ContainerItem => {
      let instance;

      if (
        typeof ContainerItem === 'object' &&
        ContainerItem instanceof Container
      ) {
        instance = ContainerItem;
      } else {
        instance = safeMap.get(ContainerItem);

        if (!instance) {
          instance = new ContainerItem();
          safeMap.set(ContainerItem, instance);
        }
      }

      instance.unsubscribe(this.onUpdate);
      instance.subscribe(this.onUpdate);

      return instance;
    });

    this.instances = instances;
    return instances;
  }

  render() {
    return (
      <StateContext.Consumer>
        {map =>
          this.props.children.apply(
            null,
            this._createInstances(map, this.props.to)
          )
        }
      </StateContext.Consumer>
    );
  }
}

export type ProviderProps = {
  inject?: Array<ContainerType>,
  children: Node
};

export function Provider(props: ProviderProps) {
  return (
    <StateContext.Consumer>
      {parentMap => {
        let childMap = new Map(parentMap);

        if (props.inject) {
          props.inject.forEach(instance => {
            childMap.set(instance.constructor, instance);
          });
        }

        return (
          <StateContext.Provider value={childMap}>
            {props.children}
          </StateContext.Provider>
        );
      }}
    </StateContext.Consumer>
  );
}
