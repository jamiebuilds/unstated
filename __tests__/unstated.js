// @flow
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider, Subscribe, Container } from '../src/unstated';

const noop = () => {};

Enzyme.configure({ adapter: new Adapter() });

class CounterContainer extends Container<{ count: number }> {
  state = { count: 0 };
  increment(amount = 1) {
    this.setState({ count: this.state.count + amount });
  }
  decrement(amount = 1) {
    this.setState({ count: this.state.count - amount });
  }
}

class AmounterContainer extends Container<{ amount: number }> {
  state = { amount: 1 };
  setAmount(amount) {
    this.setState({ amount });
  }
}

function Counter() {
  return (
    <Subscribe to={[CounterContainer]}>
      {counter => (
        <div>
          <span>{counter.state.count}</span>
          <button onClick={() => counter.decrement()}>-</button>
          <button onClick={() => counter.increment()}>+</button>
        </div>
      )}
    </Subscribe>
  );
}

function CounterWithAmount() {
  return (
    <Subscribe to={[CounterContainer, AmounterContainer]}>
      {(counter, amounter) => (
        <div>
          <span>{counter.state.count}</span>
          <button onClick={() => counter.decrement(amounter.state.amount)}>
            -
          </button>
          <button onClick={() => counter.increment(amounter.state.amount)}>
            +
          </button>
        </div>
      )}
    </Subscribe>
  );
}

function CounterWithAmountApp() {
  return (
    <Subscribe to={[AmounterContainer]}>
      {amounter => (
        <div>
          <Counter />
          <input
            type="number"
            value={amounter.state.amount}
            onChange={event => {
              amounter.setAmount(parseInt(event.currentTarget.value, 10));
            }}
          />
        </div>
      )}
    </Subscribe>
  );
}

test('component with Subscribe renders fine', () => {
  const wrapper = mount(
    <Provider>
      <Counter />
    </Provider>
  );
  expect(wrapper).toMatchSnapshot();
});

test("component can change container's state", () => {
  const wrapper = mount(
    <Provider>
      <Counter />
    </Provider>
  );
  expect(wrapper).toMatchSnapshot();
  const incButton = wrapper.find('button').last();
  incButton.simulate('click');
  incButton.simulate('click');
  expect(wrapper).toMatchSnapshot();
});

test("throws when not in a Provider's subtree", () => {
  const consoleError = console.error;
  console.error = noop;
  expect(() => mount(<Counter />)).toThrowErrorMatchingSnapshot();
  console.error = consoleError;
});

test('Subscribes share Containers instances', () => {
  const wrapper = mount(
    <Provider>
      <Counter />
      <Counter />
      <Counter />
    </Provider>
  );
  expect(wrapper).toMatchSnapshot();
  const incButton = wrapper
    .find(Counter)
    .first()
    .find('button')
    .last();
  incButton.simulate('click');
  incButton.simulate('click');
  expect(wrapper).toMatchSnapshot();
});
