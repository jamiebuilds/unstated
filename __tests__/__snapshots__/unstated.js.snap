// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribes share Containers instances 1`] = `
<Provider>
  <Consumer>
    <Provider
      value={
        Map {
          [Function] => CounterContainer {
            "_listeners": Array [
              [Function],
              [Function],
              [Function],
            ],
            "state": Object {
              "count": 0,
            },
          },
        }
      }
    >
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                0
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                0
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                0
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
    </Provider>
  </Consumer>
</Provider>
`;

exports[`Subscribes share Containers instances 2`] = `
<Provider>
  <Consumer>
    <Provider
      value={
        Map {
          [Function] => CounterContainer {
            "_listeners": Array [
              [Function],
              [Function],
              [Function],
            ],
            "state": Object {
              "count": 2,
            },
          },
        }
      }
    >
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                2
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                2
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                2
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
    </Provider>
  </Consumer>
</Provider>
`;

exports[`component can change container's state 1`] = `
<Provider>
  <Consumer>
    <Provider
      value={
        Map {
          [Function] => CounterContainer {
            "_listeners": Array [
              [Function],
            ],
            "state": Object {
              "count": 0,
            },
          },
        }
      }
    >
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                0
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
    </Provider>
  </Consumer>
</Provider>
`;

exports[`component can change container's state 2`] = `
<Provider>
  <Consumer>
    <Provider
      value={
        Map {
          [Function] => CounterContainer {
            "_listeners": Array [
              [Function],
            ],
            "state": Object {
              "count": 2,
            },
          },
        }
      }
    >
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                2
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
    </Provider>
  </Consumer>
</Provider>
`;

exports[`component with Subscribe renders fine 1`] = `
<Provider>
  <Consumer>
    <Provider
      value={
        Map {
          [Function] => CounterContainer {
            "_listeners": Array [
              [Function],
            ],
            "state": Object {
              "count": 0,
            },
          },
        }
      }
    >
      <Counter>
        <Subscribe
          to={
            Array [
              [Function],
            ]
          }
        >
          <Consumer>
            <div>
              <span>
                0
              </span>
              <button
                onClick={[Function]}
              >
                -
              </button>
              <button
                onClick={[Function]}
              >
                +
              </button>
            </div>
          </Consumer>
        </Subscribe>
      </Counter>
    </Provider>
  </Consumer>
</Provider>
`;

exports[`throws when not in a Provider's subtree 1`] = `"You must wrap your <Subscribe> components with a <Provider>"`;
